File: Day5/Day5.cpp
#include <iostream>
#include <fstream> 
#include <string> 
#include <utility> 
#include <map>

using namespace std;

pair<string, string> split(string original, string delimiter) {
    size_t index = original.find(delimiter);
    string token_1 = original.substr(0, index);
    string token_2 = original.substr(index + delimiter.length(), original.length()-1);
    return pair<string, string>(token_1, token_2);
} 

int line_intersection(ifstream& file, bool diagonals) {
    int intersections = 0;

    map<pair<int, int>, int> line_map;
    string line("");
    while(getline(file, line)) {
        pair<string, string> start_finish = split(line, " -> ");
        pair<string, string> start_str = split(start_finish.first, ",");
        pair<string, string> finish_str = split(start_finish.second, ",");

        int x1 = stoi(start_str.first);
        int y1 = stoi(start_str.second);
        int x2 = stoi(finish_str.first);
        int y2 = stoi(finish_str.second);

        int step_y = (y2-y1)/max(abs(y2-y1), 1);
        int step_x = (x2-x1)/max(abs(x2-x1), 1);

        if(x1 == x2 || y1 == y2) {
            for(int i = y1; i != y2+step_y; i+=step_y) {
                auto it = line_map.find(make_pair(x1, i)); 
                if (it == line_map.end()) {
                    line_map.insert(make_pair(make_pair(x1, i), 1));
                } else {
                    it->second++; 
                }
            }

            for(int i = x1; i != x2+step_x; i+=step_x) {
                auto it = line_map.find(make_pair(i, y1)); 
                if (it == line_map.end()) {
                    line_map.insert(make_pair(make_pair(i, y1), 1));
                } else {
                    it->second++; 
                }
            }
        } else if(diagonals) {
            for(int i = x1; i != x2+step_x; i+=step_x) {
                auto it = line_map.find(make_pair(i, y1)); 
                if (it == line_map.end()) {
                    line_map.insert(make_pair(make_pair(i, y1), 1));
                } else {
                    it->second++; 
                }
                y1+=step_y;
            }
        }
    }
    map<pair<int, int>, int>::iterator it;
    for(it = line_map.begin(); it != line_map.end(); it++) {
        if(it->second > 1) intersections++;
    }
    
    return intersections;
}

int main(int argc, char* argv[]) {
    ifstream file(argv[1]);  
    printf("Part1: %d\n", line_intersection(file, false));
    //Reset file cursor
    file.clear();
    file.seekg(0);
    //-----------------
    printf("Part2: %d\n", line_intersection(file, true));
    file.close();
    return 0;
}

File: Day5/Day5.py
import sys

def check_line_intersection(file_lines, diagonal):
    positions = {}
    count = 0
    for line in file_lines:
        coordinates = line.split(" -> ")
        x1, y1 = [int(k) for k in coordinates[0].split(",")]
        x2, y2 = [int(k) for k in coordinates[1].split(",")]
        step_y = (y2-y1)//max(abs(y2-y1), 1)
        step_x = (x2-x1)//max(abs(x2-x1), 1)
        if x1 == x2:
            for i in range(y1, y2+step_y, step_y):
                if (x1, i) not in positions.keys():
                    positions[(x1, i)] = 1
                else:
                    positions[(x1, i)] += 1
        elif y1 == y2:
            for i in range(x1, x2+step_x, step_x):
                if (i, y1) not in positions.keys():
                    positions[(i, y1)] = 1
                else:
                    positions[(i, y1)] += 1
        elif diagonal:
            while x1 != x2+step_x:
                if (x1, y1) not in positions.keys():
                    positions[(x1, y1)] = 1
                else:
                    positions[(x1, y1)] += 1
                x1 += step_x 
                y1 += step_y
    

    for key in positions.keys():
        if positions[key] > 1:
            count += 1
    return count



if __name__ == "__main__":
    file = open(sys.argv[1], "r")
    file_lines = file.readlines()
    print("Part1: ", check_line_intersection(file_lines, False))
    print("Part2: ", check_line_intersection(file_lines, True))
    file.close()
